lib/executor/constant_arrival_rate.go

```diff
@@ -300,28 +326,40 @@ func (car ConstantArrivalRate) Run(ctx context.Context, out chan<- stats.SampleC
 				int64(time.Duration(car.config.TimeUnit.Duration)),
 			)).Duration)
 
+	shownWarning := false
 	for li, gi := 0, start; ; li, gi = li+1, gi+offsets[li%len(offsets)] {
 		t := notScaledTickerPeriod*time.Duration(gi) - time.Since(startTime)
 		timer.Reset(t)
 		select {
 		case <-timer.C:
 			select {
-			case vu := <-activeVUs:
-				// ideally, we get the VU from the buffer without any issues
-				go runIteration(vu)
-			default:
-				if remainingUnplannedVUs == 0 {
-					// TODO: emit an error metric?
-					car.logger.Warningf("Insufficient VUs, reached %d active VUs and cannot allocate more", maxVUs)
-					break
-				}
-				initVU, err := car.executionState.GetUnplannedVU(maxDurationCtx, car.logger)
-				if err != nil {
-					return err
+			case vu := <-activeVUs: // ideally, we get the VU from the buffer without any issues
+				go runIteration(vu) //TODO: refactor so we dont spin up a goroutine for each iteration
+				continue
+			default: // no free VUs currently available
+			}
+
+			// Since there aren't any free VUs available, consider this iteration
+			// dropped - we aren't going to try to recover it, but
+
+			// TODO: emit a dropped_iterations metric
+
+			// We'll try to start allocating another VU in the background,
+			// non-blockingly, if we have remainingUnplannedVUs...
+			if remainingUnplannedVUs == 0 {
+				if !shownWarning {
+					car.logger.Warningf("Insufficient VUs, reached %d active VUs and cannot initialize more", maxVUs)
+					shownWarning = true
```

Comment:
Should we do something more sophisticated here? :smile: This will only show it once per executor instance, which doesn't seem that useful, and could easily be missed.

Maybe show a message once every X period, with the number of VUs that weren't initialized during that period?

Reply:
This is what https://github.com/loadimpact/k6/pull/1529 is for. Instead of bothering with custom and complicated warning logic, we can emit a metric for the iterations we have missed and just issue a warning in the cloud (and maybe, in the future, in local k6 execution as well). 

> Maybe show a message once every X period, with the number of VUs that weren't initialized during that period?

The trouble is that we can't know the number of VUs that weren't initialized, we can only know how many iterations were dropped for a period, since VU initialization times are mostly unknowable. We can add such a warning in the future, if it proves useful, but I still think the `dropped_iterations` metric will be sufficient for now.

Reply:
OK, I see. Yeah, we should expose this better in local execution, maybe showing the metric in real-time as @MStoykov suggested.